### PR TITLE
Symfony 4.4 secrets management system

### PR DESCRIPTION
**Pull request by weaverryan**
Hi!

This completes #11396

I also made some changes to the section in `configuration.rst` about environment variables to make sure that it read well with the secrets stuff.

Thanks to @jderusse for doing the vast majority of the work.

Cheers!

### CODE REVIEW

File: configuration/secrets.rst [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    with PHP 7.2. If you're using an earlier PHP version, you can
+    install the `libsodium`_ PHP extension or use the
+    `paragonie/sodium_compat`_ package.
+
+.. _secrets-generate-keys:
+
+Generate Cryptographic Keys
+---------------------------
+
+In order to encrypt and decrypt **secrets**, symfony needs **cryptographic keys**.
```
Comment (94noni) [inline configuration/secrets.rst]: Symfony, no?

File: configuration/secrets.rst [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+This will generate ``config/secrets/prod/prod.encrypt.public.php`` and
+``config/secrets/prod/prod.decrypt.private.php``.
+
+.. caution::
+
+    The ``prod.decrypt.private.php`` file is highly sensitive. Your team of developers
+    and even Continuous Integration services don't need that key. If the
+    **decryption key** has been exposed (ex-employee leaving for instance), you
+    should consider generating a new one by running:
+    ``secrets:generate-keys --rotate``.
```
Comment (94noni) [inline configuration/secrets.rst]: Need to readd —env=prod IMO as you are explaining prod logic

File: configuration/secrets.rst [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    $ php bin/console secrets:set DATABASE_PASSWORD --local
+
+If you entered ``root``, you will now see this in your ``.env.dev.local`` file:
+
+.. code-block:: bash
+
+    DATABASE_PASSWORD=root
+
+This will *override* the ``DATABASE_PASSWORD`` secret because environment variables
+always "win" over secrets.
```
Comment (94noni) [inline configuration/secrets.rst]: Takes precedence instead of win ?
But i am unsure of the writing

File: configuration/secrets.rst [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+.. code-block:: bash
+
+    # .env.test
+    DATABASE_PASSWORD="testing"
+
+Deploy secrets to production
+----------------------------
+
+Because the **decryption key** is not committed, you'll need to manually
```
Comment (94noni) [inline configuration/secrets.rst]: Maybe reword, as a reminder

Due to the fact that decryption keys should never be commited, you will need ...

Wdyt?

File: configuration.rst [hunk 5]
```diff
@@ -485,164 +485,168 @@ This example shows how to configure the database connection using an env var:
...
+you can add:
 
 .. code-block:: bash
 
     # .env
     DATABASE_URL="mysql://db_user:db_password@127.0.0.1:3306/db_name"
 
+This file should be committed to your repository and (due to that fact) should
+only contain "default" values that are good for local development. This file
+should not contain production values.
```
[the remark this answers is not in the document]
Comment (jderusse) [reply configuration.rst]: sensitive value should not be committed and therefore should not be stored in .env file which contains default public values.

I'd rather use secrets to store "pre-prodction" sensitive value and change the secret location (see section configuration) according to an env variable that define the stage.
ie.
```yaml
framework:
  secrets:
    vault_directory: '%kernel.project_dir%/config/secrets/%kernel.environment%/%app_stage'
parameters:
  app_stage: '%env(STAGE)%'
```

```bash
$ STAGE=PRE-PRODCTION APP_ENV=PROD bin/console secrets:set DATABASE_PASSWORD
```

**Review by wouterj — APPROVED** [review]
Nice work!

File: best_practices.rst [hunk 6]
```diff
@@ -87,6 +87,12 @@ application behavior.
 :ref:`Use env vars in your project <config-env-vars>` to define these options
 and create multiple ``.env`` files to :ref:`configure env vars per environment <config-dot-env>`.
 
+Use Secret for Sensitive Information
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When your application has sensitive configuration - like an API key - you should
```
[the remark this answers is not in the document]
Comment (weaverryan) [reply best_practices.rst]: I see your point. But on some systems (Heroku, SymfonyCloud), the database credentials are provided to you by your platform and are not something you need to store in a vault.

File: configuration.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (weaverryan) [reply configuration.rst]: This might be a nice example to add to the config section after this PR is merged :)

File: configuration.rst [hunk 7]
```diff
@@ -485,164 +485,168 @@ This example shows how to configure the database connection using an env var:
...
 
-    # run this command to show all the details for a specific env var:
-    $ php bin/console debug:container --env-var=FOO
+.. caution::
 
-.. versionadded:: 4.3
+    Beware that dumping the contents of the ``$_SERVER`` and ``$_ENV`` variables
+    or outputting the ``phpinfo()`` contents will display the values of the
+    environment variables, exposing sensitive information such as the database
+    credentials.
```
[the remark this answers is not in the document]
Comment (weaverryan) [reply configuration.rst]: That's true... but only true when users are setting "real" env vars (not using `.env`), right? In general, I hope we can now (with secrets) *really* push people to use that system. This warning already is starting to look misplaced to me: we should (in the document in general) be telling people to instantly use the secrets management for sensitive stuff.

**Comment by weaverryan** [comment]
This should be ready now!

File: configuration/secrets.rst [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+If you stored a ``DATABASE_PASSWORD`` secret, you can reference by:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/doctrine.yaml
+        doctrine:
+            dbal:
+                password: '%env(DATABASE_PASSWORD)%'
```
Comment (ghostika) [inline configuration/secrets.rst]: Shouldn't this be `%env(secret:DATABASE_PASSWORD)` ?

Comment (jderusse) [reply configuration/secrets.rst]: there is no `secret` env processor, the document is right here, to access secret, you should use `%env(DATABASE_PASSWORD)%`. this is explain in the paragraph above (line 104).

**Comment by weaverryan** [comment]
Thanks to @jderusse for helping with this great feature and writing most of the docs. Sorry for the delay - but I wanted to get the first draft right :)
